packages/react-reconciler/src/ReactFiberClassComponent.js

```diff
@@ -513,27 +513,52 @@ function constructClassInstance(
   let unmaskedContext = emptyContextObject;
   let context = null;
   const contextType = ctor.contextType;
-  if (typeof contextType === 'object' && contextType !== null) {
-    if (__DEV__) {
-      const isContextConsumer =
+
+  if (__DEV__) {
+    if ('contextType' in ctor) {
+      let isValid =
+        contextType !== null &&
+        contextType !== undefined &&
         contextType.$$typeof === REACT_CONTEXT_TYPE &&
-        contextType._context !== undefined;
-      if (
-        (contextType.$$typeof !== REACT_CONTEXT_TYPE || isContextConsumer) &&
-        !didWarnAboutInvalidateContextType.has(ctor)
-      ) {
+        contextType._context === undefined; // Not a <Context.Consumer>
+
+      if (!isValid && !didWarnAboutInvalidateContextType.has(ctor)) {
         didWarnAboutInvalidateContextType.add(ctor);
+
+        let addendum = '';
+        if (contextType === null) {
+          addendum = ' However, it is set to null.';
+        } else if (contextType === undefined) {
+          addendum =
+            ' However, it is set to undefined. ' +
+            'This can be caused by a typo or by mixing up named and default imports. ' +
+            'This can also happen due to a circular dependency, so ' +
+            'try moving the createContext() call to a separate file.';
+        } else if (typeof contextType !== 'object') {
+          addendum = ' However, it is set to ' + typeof contextType + '.';
```

Comment:
I think we can add "a" here to make it a little clearer

```js
addendum = ' However, it is set to a ' + typeof contextType + '.';
```
That way it reports "set to a string" instead of "set to string", which sounds better. Same goes for numbers, symbols, and functions.